pkgs/_macro_host/lib/src/package_config.dart

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:dart_model/dart_model.dart';
+import 'package:package_config/package_config.dart';
+
+/// Reads a package config to determine information about macros.
+class MacroPackageConfig {
+  final PackageConfig packageConfig;
+
+  MacroPackageConfig({required this.packageConfig});
+
+  factory MacroPackageConfig.readFromUri(Uri uri) => MacroPackageConfig(
+      packageConfig:
+          PackageConfig.parseBytes(File.fromUri(uri).readAsBytesSync(), uri));
+
+  /// Checks whether [name] is a macro annotation.
+  ///
+  /// If so, returns the qualified name of the macro implementation.
+  ///
+  /// If not, returns `null`.
+  ///
+  /// This is a placeholder implementation until `language/3728` is
+  /// implemented. It expects macros to be marked by a comment in the
+  /// annotation package `pubspec.yaml` that looks like this:
+  ///
+  /// ```
+  /// # macro <annotation name> <implementation name>
+  /// ```
+  ///
+  /// For example:
+  ///
+  /// ```
+  /// # macro lib/declare_x_macro.dart#DeclareX package:_test_macros/declare_x_macro.dart#DeclareXImplementation
+  /// ```
+  QualifiedName? lookupMacroImplementation(QualifiedName name) {
+    var packageName = name.uri;
+    if (packageName.startsWith('dart:')) return null;
+    if (packageName.startsWith('package:') && packageName.contains('/')) {
+      packageName = packageName.substring('package:'.length);
+      packageName = packageName.substring(0, packageName.indexOf('/'));
+    } else {
+      // TODO(davidmorgan): support macros outside lib dirs.
+      throw ArgumentError('Name must start "package:" and have a path: $name');
+    }
+    final libraryPath =
+        'lib/${name.string.substring(name.uri.indexOf('/') + 1)}';
+
+    final matchingPackage = packageConfig.packages
+        .where((p) => p.name == packageName)
+        .toList()
+        .singleOrNull;
```

Comment:
I am sure it doesn't matter all that much, but creating a closure + iterable + list here does bug me :).

Luckily, `PackageConfig` implements the `[]` operator :).

```suggestion
    final matchingPackage = packageConfig[packageName];
```

Reply:
Done.